src/stream/src/executor/over_window/general.rs

```diff
@@ -0,0 +1,944 @@
+                                    old_row: *real_old_row,
+                                };
+                            }
+                            _ => panic!("inconsistent changes in input chunk"),
+                        }
+                    } else {
+                        changes_merged.insert(pk, record);
+                    }
+                }
+                Record::Update { old_row, new_row } => {
+                    let pk = DefaultOrdered(this.get_input_pk(old_row));
+                    if let Some(prev_change) = changes_merged.get_mut(&pk) {
+                        match prev_change {
+                            Record::Insert { .. } => {
+                                *prev_change = Record::Insert { new_row };
+                            }
+                            Record::Update {
+                                old_row: real_old_row,
+                                ..
+                            } => {
+                                *prev_change = Record::Update {
+                                    old_row: *real_old_row,
+                                    new_row,
+                                };
+                            }
+                            _ => panic!("inconsistent changes in input chunk"),
+                        }
+                    } else {
+                        changes_merged.insert(pk, record);
+                    }
+                }
+            }
+        }
+        changes_merged.into_values()
+    }
+
+    #[try_stream(ok = StreamChunk, error = StreamExecutorError)]
+    async fn apply_chunk<'a>(
+        this: &'a mut ExecutorInner<S>,
+        vars: &'a mut ExecutionVars<S>,
+        chunk: StreamChunk,
+    ) {
+        // `partition key` => `Delta`.
+        let mut deltas: BTreeMap<DefaultOrdered<OwnedRow>, Delta> = BTreeMap::new();
+        // `input pk` of update records of which the `partition key` or `order key` is changed.
+        let mut key_change_updated_pks = HashSet::new();
+
+        // Collect changes for each partition.
+        for record in Self::merge_changes_in_chunk(this, &chunk) {
+            match record {
+                Record::Insert { new_row } => {
+                    let part_key = this.get_partition_key(new_row).into();
+                    let part_delta = deltas.entry(part_key).or_insert(Delta::new());
+                    part_delta.insert(
+                        this.row_to_state_key(new_row)?,
+                        Change::Insert(new_row.into_owned_row()),
+                    );
+                }
+                Record::Delete { old_row } => {
+                    let part_key = this.get_partition_key(old_row).into();
+                    let part_delta = deltas.entry(part_key).or_insert(Delta::new());
+                    part_delta.insert(this.row_to_state_key(old_row)?, Change::Delete);
+                }
+                Record::Update { old_row, new_row } => {
```

Comment:
Can we simply treat this as a combination of `Delete` and `Insert`?

Reply:
Because OverWindow produces one and exactly one output row for each input row, it seems a little bit strange if an `Update` produces a `Delete` + `Insert`, especially for order-change updates that are not crossing partition. Also, maintaining input information with just a small cost seems reasonable.

Reply:
Another reason: For window function calls with different `PARTITION BY` and `ORDER BY`, the plan will be a chain of OverWindow, if we simply split `Update` into `Delete` + `Insert`, the output `Insert` may be passed to downstream before `Delete` (if they happened to be in different chunk cuz the `chunk_builder` reaching `chunk_size`), this can cause significant chaos.

Reply:
> Another reason: For window function calls with different `PARTITION BY` and `ORDER BY`, the plan will be a chain of OverWindows, if we simply split `Update` into `Delete` + `Insert`, the output `Insert` may be passed to downstream before `Delete` (if they happened to be in different chunks because of the `chunk_builder` reaching `chunk_size`), this can cause significant chaos.

Sounds like a potential problem? Logically, `update` events can always be replaced with `delete`+`insert`, for example, a user can deliberately do this by using `delete` + `insert` statements.

Reply:
Yes. This is resolved in [`c1bbe38` (#10525)](https://github.com/risingwavelabs/risingwave/pull/10525/commits/c1bbe383b5ff7a9895030b51bc825c1a6f1a837f#diff-4976b08fc158f56a8740c7f3a753e2f235e157d76766b077eaeab1dbb722bb89), by split partition-change `Update` to `Delete` + `Insert` to downstream while still merge order-change `Update`s (which is required for correctness).

Reply:
To clarify, let's assume here we only have `insert` and `delete` events. 

Assuming the case that an `insert` have some dependency with the `delete` event, and they belong to 2 different partition. For example, the `delete` deletes the row just being `insert`ed, or the `insert` inserts a row with the same PK of previously `delete`d row.

In that case, we should never reorder this pair of `insert` and `delete` event, because the resulted change may still have some kind of dependency, and reordering them will break the dependency.

Reply:
In that case, the `delete`-`insert` pairs will merged by `merge_changes_in_chunk` at the very beginning.